This Lab is based on the Text mining  and NLP Course

## Lab 1.  Data analysis with pandas




# Table of Contents
<ol start="0">
<li> Learning Goals </li>
<li> Loading and Cleaning with Pandas</li>
<li> Parsing and Completing the Dataframe  </li>
<li> Grouping </li>
</ol>

## Learning Goals

About 6000 odd "best books" were fetched and parsed from [Goodreads](https://www.goodreads.com). The "bestness" of these books came from a proprietary formula used by Goodreads and published as a list on their web site.

We parsed the page for each book and saved data from all these pages in a tabular format as a CSV file. In this lab we'll clean and further parse the data.  We'll then do some exploratory data analysis to answer questions about these best books and popular genres.  


By the end of this lab, you should be able to:

- Load and systematically address missing values, for example, by removing observations associated with these values.
- Parse columns in the dataframe to create new dataframe columns.
- Use groupby to aggregate data on a particular feature column, such as author.

Explanation:

Step 1: Loading and Handling Missing Data


When we load the dataset, some information might be missing. These missing values are marked as NaN (Not a Number).
We'll first look at these missing values, and then decide if we should remove those rows or fill in the gaps with something else.

Step 2: Parsing Data Columns


Sometimes, the data we have can be split into more detailed parts. For example, if we have a column that combines an author's full name, we might split it into "First Name" and "Last Name."
This process is called parsing, and it helps us get more specific information from our data.

Step 3: Grouping and Aggregating Data


Grouping data means organizing it based on a shared feature. For example, we can group all books by the same author.
After grouping, we can summarize the data to answer questions like, "Which author has the most books on the list?" or "Which genre is most popular?"

### Basic EDA workflow


The basic workflow is as follows:

1. **Build** a DataFrame from the data (ideally, put all data in this object)
2. **Clean** the DataFrame. It should have the following properties:
    - Each row describes a single object
    - Each column describes a property of that object
    - Columns are numeric whenever appropriate
    - Columns contain atomic properties that cannot be further decomposed
3. Explore **global properties**. Use histograms, scatter plots, and aggregation functions to summarize the data.
4. Explore **group properties**. Use groupby and small multiples to compare subsets of the data.

This process transforms your data into a format which is easier to work with, gives you a basic overview of the data's properties, and likely generates several questions for you to followup in subsequent analysis.

## Part 1: Loading and Cleaning with Pandas
Read in the `goodreads.csv` file, examine the data, and do any necessary data cleaning.

Here is a description of the columns (in order) present in this csv file:

```
rating: the average rating on a 1-5 scale achieved by the book
review_count: the number of Goodreads users who reviewed this book
isbn: the ISBN code for the book
booktype: an internal Goodreads identifier for the book
author_url: the Goodreads (relative) URL for the author of the book
year: the year the book was published
genre_urls: a string with '|' separated relative URLS of Goodreads genre pages
dir: a directory identifier internal to the scraping code
rating_count: the number of ratings for this book (this is different from the number of reviews)
name: the name of the book
```

Let us see what issues we find with the data and resolve them.  



----




After loading appropriate libraries


In [1]:
%matplotlib inline
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
pd.set_option('display.width', 500)
pd.set_option('display.max_columns', 100)

### Cleaning: Reading in the data
We read in and clean the data from `goodreads.csv`.

In [3]:
#Read the data into a dataframe
df = pd.read_csv("goodreads.csv")

#Examine the first few rows of the dataframe
df.head()

4.40   136455  0439023483  good_reads:book https://www.goodreads.com/author/show/153394.Suzanne_Collins    2008 /genres/young-adult|/genres/science-fiction|/genres/dystopia|/genres/fantasy|/genres/science-fiction|/genres/romance|/genres/adventure|/genres/book-club|/genres/young-adult|/genres/teen|/genres/apocalyptic|/genres/post-apocalyptic|/genres/action                dir01/2767052-the-hunger-games.html    2958974            The Hunger Games (The Hunger Games, #1)
0  4.41  16648.0  0439358078  good_reads:book  https://www.goodreads.com/author/show/1077326....            2003.0  /genres/fantasy|/genres/young-adult|/genres/fi...                                                                                                                                                                                                     dir01/2.Harry_Potter_and_the_Order_of_the_Phoe...  1284478.0  Harry Potter and the Order of the Phoenix (Har...
1  3.56  85746.0  0316015849  good_reads:book  https://www.goodreads.com/author/show/941441.S...            2005.0  /genres/young-adult|/genres/fantasy|/genres/ro...                                                                                                                                                                                                                             dir01/41865.Twilight.html  2579564.0                            Twilight (Twilight, #1)
2  4.23  47906.0  0061120081  good_reads:book  https://www.goodreads.com/author/show/1825.Har...            1960.0  /genres/classics|/genres/fiction|/genres/histo...                                                                                                                                                                                                                 dir01/2657.To_Kill_a_Mockingbird.html  2078123.0                              To Kill a Mockingbird
3  4.23  34772.0  0679783261  good_reads:book  https://www.goodreads.com/author/show/1265.Jan...            1813.0  /genres/classics|/genres/fiction|/genres/roman...                                                                                                                                                                                                                   dir01/1885.Pride_and_Prejudice.html  1388992.0                                Pride and Prejudice
4  4.25  12363.0  0446675539  good_reads:book  https://www.goodreads.com/author/show/11081.Ma...            1936.0  /genres/classics|/genres/historical-fiction|/g...                                                                                                                                                                                                                   dir01/18405.Gone_with_the_Wind.html   645470.0                                 Gone with the Wind

Oh dear. That does not quite seem to be right. We are missing the column names. We need to add these in! But what are they?

Here is a list of them in order:

`["rating", 'review_count', 'isbn', 'booktype','author_url', 'year', 'genre_urls', 'dir','rating_count', 'name']`

<div class="exercise"><b>Exercise</b></div>
Use these to load the dataframe properly! And then "head" the dataframe... (you will need to look at the read_csv docs)


In [4]:
df.columns=["rating", 'review_count', 'isbn', 'booktype','author_url', 'year', 'genre_urls', 'dir','rating_count', 'name']
df.head()

rating  review_count        isbn         booktype                                         author_url    year                                         genre_urls                                                dir  rating_count                                               name
0    4.41       16648.0  0439358078  good_reads:book  https://www.goodreads.com/author/show/1077326....  2003.0  /genres/fantasy|/genres/young-adult|/genres/fi...  dir01/2.Harry_Potter_and_the_Order_of_the_Phoe...     1284478.0  Harry Potter and the Order of the Phoenix (Har...
1    3.56       85746.0  0316015849  good_reads:book  https://www.goodreads.com/author/show/941441.S...  2005.0  /genres/young-adult|/genres/fantasy|/genres/ro...                          dir01/41865.Twilight.html     2579564.0                            Twilight (Twilight, #1)
2    4.23       47906.0  0061120081  good_reads:book  https://www.goodreads.com/author/show/1825.Har...  1960.0  /genres/classics|/genres/fiction|/genres/histo...              dir01/2657.To_Kill_a_Mockingbird.html     2078123.0                              To Kill a Mockingbird
3    4.23       34772.0  0679783261  good_reads:book  https://www.goodreads.com/author/show/1265.Jan...  1813.0  /genres/classics|/genres/fiction|/genres/roman...                dir01/1885.Pride_and_Prejudice.html     1388992.0                                Pride and Prejudice
4    4.25       12363.0  0446675539  good_reads:book  https://www.goodreads.com/author/show/11081.Ma...  1936.0  /genres/classics|/genres/historical-fiction|/g...                dir01/18405.Gone_with_the_Wind.html      645470.0                                 Gone with the Wind

### Cleaning: Examing the dataframe - quick checks

We should examine the dataframe to get a overall sense of the content.

<div class="exercise"><b>Exercise</b></div>
Lets check the types of the columns. What do you find?

In [5]:
datatypes = df.dtypes
datatypes

rating          float64
review_count    float64
isbn             object
booktype         object
author_url       object
year            float64
genre_urls       object
dir              object
rating_count    float64
name             object
dtype: object

The DataFrame contains:
- 4 columns with float64 type: (rating, review_count, year, rating_count) are numeric.
- 6 columns with object type: (isbn, booktype, author_url, genre_urls, dir, name) likely contain text or mixed data.



In [6]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5999 entries, 0 to 5998
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   rating        5997 non-null   float64
 1   review_count  5997 non-null   float64
 2   isbn          5522 non-null   object 
 3   booktype      5997 non-null   object 
 4   author_url    5997 non-null   object 
 5   year          5992 non-null   float64
 6   genre_urls    5937 non-null   object 
 7   dir           5999 non-null   object 
 8   rating_count  5997 non-null   float64
 9   name          5997 non-null   object 
dtypes: float64(4), object(6)
memory usage: 468.8+ KB
None


The DataFrame has 5,999 rows and 10 columns.
Missing values are found in:
- isbn (477 missing)
- genre_urls (62 missing)
- year (7 missing).

Data cleaning is needed for missing values and format inconsistencies.

There are a couple more quick sanity checks to perform on the dataframe.

In [7]:
print(df.shape)
df.columns

(5999, 10)


Index(['rating', 'review_count', 'isbn', 'booktype', 'author_url', 'year', 'genre_urls', 'dir', 'rating_count', 'name'], dtype='object')

### Cleaning: Examining the dataframe - a deeper look

Beyond performing checking some quick general properties of the data frame and looking at the first $n$ rows, we can dig a bit deeper into the values being stored. If you haven't already, check to see if there are any missing values in the data frame.

Let's see for a column which seemed OK to us.

In [8]:
#Get a sense of how many missing values there are in the dataframe. Do it for all columns
np.sum([df.rating.isnull()])

2

In [9]:
#Try to locate where the missing values occur
df[df.rating.isnull()]

rating  review_count isbn booktype author_url  year genre_urls                              dir  rating_count name
3642     NaN           NaN  NaN      NaN        NaN   NaN        NaN  dir37/9658936-harry-potter.html           NaN  NaN
5281     NaN           NaN  NaN      NaN        NaN   NaN        NaN     dir53/113138.The_Winner.html           NaN  NaN

How does `pandas` or `numpy` handle missing values when we try to compute with data sets that include them?

We'll now check if any of the other suspicious columns have missing values.  Let's look at `year` and `review_count` first.

One thing you can do is to try and convert to the type you expect the column to be. If something goes wrong, it likely means your data are bad.

Lets test for missing data:

In [10]:
df[df.year.isnull()]

rating  review_count        isbn         booktype                                         author_url  year                                         genre_urls                                                dir  rating_count                                               name
2441    4.23         526.0         NaN  good_reads:book  https://www.goodreads.com/author/show/623606.A...   NaN  /genres/religion|/genres/islam|/genres/non-fic...                       dir25/1301625.La_Tahzan.html        4134.0                                          La Tahzan
2868    4.61           2.0         NaN  good_reads:book  https://www.goodreads.com/author/show/8182217....   NaN                                                NaN  dir29/22031070-my-death-experiences---a-preach...          23.0  My Death Experiences - A Preacher’s 18 Apoca...
3642     NaN           NaN         NaN              NaN                                                NaN   NaN                                                NaN                    dir37/9658936-harry-potter.html           NaN                                                NaN
5281     NaN           NaN         NaN              NaN                                                NaN   NaN                                                NaN                       dir53/113138.The_Winner.html           NaN                                                NaN
5571    3.71          35.0  8423336603  good_reads:book  https://www.goodreads.com/author/show/285658.E...   NaN                                    /genres/fiction  dir56/890680._rase_una_vez_el_amor_pero_tuve_q...         403.0  Érase una vez el amor pero tuve que matarlo. ...
5657    4.32          44.0         NaN  good_reads:book  https://www.goodreads.com/author/show/25307.Ro...   NaN  /genres/fantasy|/genres/fantasy|/genres/epic-f...  dir57/5533041-assassin-s-apprentice-royal-assa...        3850.0  Assassin's Apprentice / Royal Assassin (Farsee...
5682    4.56         204.0         NaN  good_reads:book  https://www.goodreads.com/author/show/3097905....   NaN  /genres/fantasy|/genres/young-adult|/genres/ro...                  dir57/12474623-tiger-s-dream.html         895.0                 Tiger's Dream (The Tiger Saga, #5)

### Cleaning: Dealing with Missing Values
How should we interpret 'missing' or 'invalid' values in the data (hint: look at where these values occur)? One approach is to simply exclude them from the dataframe. Is this appropriate for all 'missing' or 'invalid' values?

In [11]:
#Treat the missing or invalid values in your dataframe

df = df[df.year.notnull()]

Ok so we have done some cleaning. What do things look like now? Notice the float has not yet changed.

In [12]:
df.dtypes

rating          float64
review_count    float64
isbn             object
booktype         object
author_url       object
year            float64
genre_urls       object
dir              object
rating_count    float64
name             object
dtype: object

In [13]:
print(np.sum(df.year.isnull()))
print(np.sum(df.rating_count.isnull()))
print(np.sum(df.review_count.isnull()))
# We removed seven rows
df.shape

0
0
0


(5992, 10)

*Suspect observations for rating and rating_count were removed as well! *

<div class="exercise"><b>Exercise</b></div>

Ok so lets fix those types. Convert them to ints. If the type conversion fails, we now know we have further problems.

In [14]:
df['year'] = df['year'].astype(int)

Once you do this, we seem to be good on these columns (no errors in conversion). Lets look:

In [15]:
df.dtypes

rating          float64
review_count    float64
isbn             object
booktype         object
author_url       object
year              int64
genre_urls       object
dir              object
rating_count    float64
name             object
dtype: object

Sweet!

Some of the other colums that should be strings have NaN.

In [16]:
df.loc[df.genre_urls.isnull(), 'genre_urls']=""
df.loc[df.isbn.isnull(), 'isbn']=""

##  Part 2: Parsing and Completing the Data Frame

We will parse the `author` column from the author_url and `genres` column from the genre_urls. Keep the `genres` column as a string separated by '|'.

We will use panda's `map` to assign new columns to the dataframe.  

Examine an example `author_url` and reason about which sequence of string operations must be performed in order to isolate the author's name.

In [17]:
#Get the first author_url
test_string = df.author_url[0]
test_string

'https://www.goodreads.com/author/show/1077326.J_K_Rowling'

In [18]:
#Test out some string operations to isolate the author name

test_string.split('/')[-1].split('.')[1:][0]

'J_K_Rowling'

<div class="exercise"><b>Exercise</b></div>

Lets wrap the above code into a function which we will then use

In [19]:
# Write a function that accepts an author url and returns the author's name based on your experimentation above

def get_author(url):
    # Split the URL by '/' and take the last part
    last_part = url.split('/')[-1]
    # Split the last part by '.' and extract the author's name
    if '.' in last_part:
        return last_part.split('.')[0]  # Extract the name before the first '.'
    return last_part  # Return the last part if no '.' is present


In [20]:
#Apply the get_author function to the 'author_url' column using '.map'
#and add a new column 'author' to store the names
df['author'] = df.author_url.map(get_author)
df.author[0:5]

0    1077326
1     941441
2       1825
3       1265
4      11081
Name: author, dtype: object

<div class="exercise"><b>Exercise</b></div>

Now parse out the genres from `genre_url`.  

This is a little more complicated because there be more than one genre.


In [21]:
df.genre_urls.head()

0    /genres/fantasy|/genres/young-adult|/genres/fi...
1    /genres/young-adult|/genres/fantasy|/genres/ro...
2    /genres/classics|/genres/fiction|/genres/histo...
3    /genres/classics|/genres/fiction|/genres/roman...
4    /genres/classics|/genres/historical-fiction|/g...
Name: genre_urls, dtype: object

In [23]:
#Test out some string operations to isolate the genre name
test_genre_string=df.genre_urls[0]

# Split the string into individual genres
genres=test_genre_string.strip().split('|')

# Iterate through each genre and print the last part of each URL
for e in genres:
    print(e.split('/')[-1])
    "|".join(genres)

fantasy
young-adult
fiction
fantasy
magic
childrens
adventure
science-fiction-fantasy


<div class="exercise"><b>Exercise</b></div>

Write a function that accepts a genre url and returns the genre name based on your experimentation above



In [24]:
def split_and_join_genres(genre_url):
    # Split the genre_url into individual genre URLs
    genres = genre_url.strip().split('|')

    # List to store the genre names
    genre_names = []

    # Loop through each genre and extract the genre name
    for e in genres:
        genre_name = e.split('/')[-1]  # Get the last part of the URL (genre name)
        genre_names.append(genre_name)

    # Join the genre names back into a string with '|' separator and return
    return "|".join(genre_names)


Test your function

In [25]:
split_and_join_genres("/genres/young-adult|/genres/science-fiction")

'young-adult|science-fiction'

In [26]:
split_and_join_genres("")

''

<div class="exercise"><b>Exercise</b></div>

Use map again to create a new "genres" column

In [27]:
df['genres']=df.genre_urls.map(split_and_join_genres)
df.head()

rating  review_count        isbn         booktype                                         author_url  year                                         genre_urls                                                dir  rating_count                                               name   author                                             genres
0    4.41       16648.0  0439358078  good_reads:book  https://www.goodreads.com/author/show/1077326....  2003  /genres/fantasy|/genres/young-adult|/genres/fi...  dir01/2.Harry_Potter_and_the_Order_of_the_Phoe...     1284478.0  Harry Potter and the Order of the Phoenix (Har...  1077326  fantasy|young-adult|fiction|fantasy|magic|chil...
1    3.56       85746.0  0316015849  good_reads:book  https://www.goodreads.com/author/show/941441.S...  2005  /genres/young-adult|/genres/fantasy|/genres/ro...                          dir01/41865.Twilight.html     2579564.0                            Twilight (Twilight, #1)   941441  young-adult|fantasy|romance|paranormal|vampire...
2    4.23       47906.0  0061120081  good_reads:book  https://www.goodreads.com/author/show/1825.Har...  1960  /genres/classics|/genres/fiction|/genres/histo...              dir01/2657.To_Kill_a_Mockingbird.html     2078123.0                              To Kill a Mockingbird     1825  classics|fiction|historical-fiction|academic|s...
3    4.23       34772.0  0679783261  good_reads:book  https://www.goodreads.com/author/show/1265.Jan...  1813  /genres/classics|/genres/fiction|/genres/roman...                dir01/1885.Pride_and_Prejudice.html     1388992.0                                Pride and Prejudice     1265  classics|fiction|romance|historical-fiction|li...
4    4.25       12363.0  0446675539  good_reads:book  https://www.goodreads.com/author/show/11081.Ma...  1936  /genres/classics|/genres/historical-fiction|/g...                dir01/18405.Gone_with_the_Wind.html      645470.0                                 Gone with the Wind    11081  classics|historical-fiction|fiction|romance|li...

Finally, let's pick an author at random so we can see the results of the transformations.  Scroll to see the `author` and `genre` columns that we added to the dataframe.

In [28]:
df[df.author == "Marguerite_Yourcenar"]

Empty DataFrame
Columns: [rating, review_count, isbn, booktype, author_url, year, genre_urls, dir, rating_count, name, author, genres]
Index: []

Let us delete the `genre_urls` column.

In [29]:
del df['genre_urls']

And then save the dataframe out!

In [31]:
df.to_csv("cleaned-goodreads.csv", index=False, header=True)

---

## Part 3: Grouping

It appears that some books were written in negative years!  Print out the observations that correspond to negative years.  What do you notice about these books?  

In [32]:
df[df['year'] <= 0]

rating  review_count        isbn         booktype                                         author_url  year                                                dir  rating_count                                  name   author                                             genres
46      3.68        5785.0  0143039954  good_reads:book    https://www.goodreads.com/author/show/903.Homer  -800                        dir01/1381.The_Odyssey.html      560248.0                           The Odyssey      903  classics|fiction|poetry|fantasy|mythology|acad...
245     4.01         365.0  0147712556  good_reads:book    https://www.goodreads.com/author/show/903.Homer  -800              dir03/1375.The_Iliad_The_Odyssey.html       35123.0                 The Iliad/The Odyssey      903  classics|fantasy|mythology|fantasy|academic|sc...
454     3.85        1499.0  0140449140  good_reads:book    https://www.goodreads.com/author/show/879.Plato  -380                      dir05/30289.The_Republic.html       82022.0                          The Republic      879  philosophy|classics|non-fiction|politics|histo...
595     3.77        1240.0  0679729526  good_reads:book   https://www.goodreads.com/author/show/919.Virgil   -29                        dir06/12914.The_Aeneid.html       60308.0                            The Aeneid      919  classics|poetry|fiction|fantasy|mythology|acad...
628     3.64        1231.0  1580495931  good_reads:book  https://www.goodreads.com/author/show/1002.Sop...  -429                        dir07/1554.Oedipus_Rex.html       93192.0                           Oedipus Rex     1002  classics|plays|drama|fiction|academic|school|l...
673     3.92        3559.0  1590302257  good_reads:book  https://www.goodreads.com/author/show/1771.Sun...  -512                    dir07/10534.The_Art_of_War.html      114619.0                        The Art of War     1771  non-fiction|politics|classics|literature|psych...
745     4.06        1087.0  0140449183  good_reads:book  https://www.goodreads.com/author/show/5158478....  -500                 dir08/99944.The_Bhagavad_Gita.html       31634.0                     The Bhagavad Gita  5158478  classics|spirituality|religion|hinduism|fantas...
776     3.52        1038.0  1580493882  good_reads:book  https://www.goodreads.com/author/show/1002.Sop...  -442                           dir08/7728.Antigone.html       49084.0                              Antigone     1002  drama|fiction|classics|academic|read-for-schoo...
1232    3.94         704.0  015602764X  good_reads:book  https://www.goodreads.com/author/show/1002.Sop...  -400                  dir13/1540.The_Oedipus_Cycle.html       36008.0                     The Oedipus Cycle     1002  classics|plays|drama|fiction|literature|academ...
1396    4.03         890.0  0192840509  good_reads:book  https://www.goodreads.com/author/show/12452.Aesop  -560                    dir14/21348.Aesop_s_Fables.html       71259.0                        Aesop's Fables    12452  classics|childrens|literature|fantasy|fairy-ta...
1397    3.60        1644.0  0141026286  good_reads:book  https://www.goodreads.com/author/show/5158478.... -1500             dir14/19351.The_Epic_of_Gilgamesh.html       42026.0                 The Epic of Gilgamesh  5158478  religion|literature|ancient|academic|read-for-...
1427    3.80         539.0  0486275485  good_reads:book  https://www.goodreads.com/author/show/973.Euri...  -431                            dir15/752900.Medea.html       29858.0                                 Medea      973  classics|plays|drama|fiction|fantasy|mythology...
1814    3.96         493.0  0140443339  good_reads:book  https://www.goodreads.com/author/show/990.Aesc...  -458                       dir19/1519.The_Oresteia.html       18729.0                          The Oresteia      990  classics|plays|drama|fiction|fantasy|mythology...
1881    4.02         377.0  0872205541  good_reads:book    https://www.goodreads.com/author/show/879.Plato  -400   dir19/22632.The_Trial_and_D

We can determine the "best book" by year! For this we use Pandas groupby. Groupby allows grouping a dataframe by any (usually categorical) variable.

In [33]:
dfgb_author = df.groupby('author')
type(dfgb_author)

pandas.core.groupby.generic.DataFrameGroupBy

Perhaps we want the number of books each author wrote

In [34]:
dfgb_author.count()

rating  review_count  isbn  booktype  author_url  year  dir  rating_count  name  genres
author                                                                                         
10003        4             4     4         4           4     4    4             4     4       4
10006        1             1     1         1           1     1    1             1     1       1
10015        4             4     4         4           4     4    4             4     4       4
10017        3             3     3         3           3     3    3             3     3       3
1002         3             3     3         3           3     3    3             3     3       3
...        ...           ...   ...       ...         ...   ...  ...           ...   ...     ...
9972         1             1     1         1           1     1    1             1     1       1
99849        1             1     1         1           1     1    1             1     1       1
9987         7             7     7         7           7     7    7             7     7       7
9993         1             1     1         1           1     1    1             1     1       1
9996         1             1     1         1           1     1    1             1     1       1

[2675 rows x 10 columns]

Lots of useless info there. One column should suffice

In [35]:
dfgb_author['author'].count()

author
10003    4
10006    1
10015    4
10017    3
1002     3
        ..
9972     1
99849    1
9987     7
9993     1
9996     1
Name: author, Length: 2675, dtype: int64

Perhaps you want more detailed info...

In [36]:
dfgb_author[['rating', 'rating_count', 'review_count', 'year']].describe()

rating                                                       rating_count                                                                               review_count                                                                  year                                                                  
        count      mean       std   min    25%    50%     75%   max        count          mean            std      min       25%      50%        75%       max        count    mean          std     min      25%     50%      75%      max count         mean        std     min      25%     50%      75%     max
author                                                                                                                                                                                                                                                                                                             
10003     4.0  3.907500  0.079739  3.82  3.850  3.915  3.9725  3.98          4.0  96815.000000  101516.391034  22628.0  46776.50  58900.5  108939.00  246831.0          4.0  6178.5  4872.144394  2143.0  3055.75  4769.0  7891.75  13033.0   4.0  2003.750000   4.991660  1999.0  1999.75  2003.5  2007.50  2009.0
10006     1.0  4.210000       NaN  4.21  4.210  4.210  4.2100  4.21          1.0   9199.000000            NaN   9199.0   9199.00   9199.0    9199.00    9199.0          1.0  1263.0          NaN  1263.0  1263.00  1263.0  1263.00   1263.0   1.0  2007.000000        NaN  2007.0  2007.00  2007.0  2007.00  2007.0
10015     4.0  3.842500  0.112361  3.71  3.770  3.855  3.9275  3.95          4.0  40440.000000   33471.746364  15515.0  18187.25  29021.0   51273.75   88203.0          4.0  5087.0  5552.208209  1399.0  1876.00  2826.0  6037.00  13297.0   4.0  2005.000000   7.615773  1995.0  2001.75  2006.0  2009.25  2013.0
10017     3.0  3.756667  0.049329  3.70  3.740  3.780  3.7850  3.79          3.0   7526.000000    6766.989803   3147.0   3629.00   4111.0    9715.50   15320.0          3.0   222.0   180.144387    94.0   119.00   144.0   286.00    428.0   3.0  1923.333333   9.073772  1913.0  1920.00  1927.0  1928.50  1930.0
1002      3.0  3.700000  0.216333  3.52  3.580  3.640  3.7900  3.94          3.0  59428.000000   29962.496825  36008.0  42546.00  49084.0   71138.00   93192.0          3.0   991.0   266.625205   704.0   871.00  1038.0  1134.50   1231.0   3.0  -423.666667  21.501938  -442.0  -435.50  -429.0  -414.50  -400.0
...       ...       ...       ...   ...    ...    ...     ...   ...          ...           ...            ...      ...       ...      ...        ...       ...          ...     ...          ...     ...      ...     ...      ...      ...   ...          ...        ...     ...      ...     ...      ...     ...
9972      1.0  3.540000       NaN  3.54  3.540  3.540  3.5400  3.54          1.0  30234.000000            NaN  30234.0  30234.00  30234.0   30234.00   30234.0          1.0  1832.0          NaN  1832.0  1832.00  1832.0  1832.00   1832.0   1.0  2002.000000        NaN  2002.0  2002.00  2002.0  2002.00  2002.0
99849     1.0  3.560000       NaN  3.56  3.560  3.560  3.5600  3.56          1.0  23973.000000            NaN  23973.0  23973.00  23973.0   23973.00   23973.0          1.0  1400.0          NaN  1400.0  1400.00  1400.0  1400.00   1400.0   1.0  1943.000000        NaN  1943.0  1943.00  1943.0  1943.00  1943.0
9987      7.0  3.718571  0.251358  3.25  3.655  3.800  3.8250  4.02          7.0  91456.142857  103847.313339  15219.0  40346.50  55234.0   85026.50  318994.0          7.0  4534.0  4396.255073  1346.0  2272.50  3369.0  4154.50  14169.0   7.0  2002.571429   7.322503  1987.0  2002.00  2005.0  2007.00  2008.0
9993      1.0  3.820000       NaN  3.82  3.820  3.820  3.8200  3.82          1.0  15135.000000            NaN  15135.0  15135.00  15135.0   15135.00   15135.0          1.0  3034.0          NaN  3034.0  3034.00  3034.0  3034.00   3034.0   1.0  2001.000000        NaN  2001.0  2001.00  2001.0  2001.00  2001.0
999

You can also access a `groupby` dictionary style.

In [37]:
ratingdict = {}
for author, subset in dfgb_author:
    ratingdict[author] = (subset['rating'].mean(), subset['rating'].std())
ratingdict

{'10003': (3.9075, 0.07973915809270471),
 '10006': (4.21, nan),
 '10015': (3.8425000000000002, 0.11236102527122122),
 '10017': (3.7566666666666664, 0.04932882862316234),
 '1002': (3.6999999999999997, 0.2163330765278393),
 '10022': (4.07, nan),
 '100326': (4.26, nan),
 '10039': (3.9025, 0.12257650672131269),
 '10074': (4.02, nan),
 '100752': (4.04, 0.18520259177452125),
 '10089': (3.92, nan),
 '10105': (3.79, nan),
 '10108': (3.7750000000000004, 0.049497474683058214),
 '1011': (3.82, nan),
 '101823': (4.34, nan),
 '1020792': (4.293333333333333, 0.1365039681962884),
 '102203': (4.41, nan),
 '1023510': (4.247142857142856, 0.07587583842793977),
 '1025097': (3.97, 0.04301162633521312),
 '10263': (4.29, nan),
 '10264': (3.87, 0.14142135623730964),
 '10289': (4.06, 0.09643650760992951),
 '10317': (3.813333333333334, 0.13428824718989124),
 '10330': (3.9699999999999998, 0.04242640687119289),
 '1033445': (3.43, nan),
 '10356': (4.03, nan),
 '10366': (4.108, 0.056302753041037115),
 '1036615': (3.

<div class="exercise"><b>Exercise</b></div>

Lets get the best-rated book(s) for every year in our dataframe.

In [38]:
#Using .groupby, we can divide the dataframe into subsets by the values of 'year'.
#We can then iterate over these subsets

dfgb_year = df.groupby('year') # Group the dataframe by 'year'

# Iterate over each year and its corresponding subset
for year, subset in dfgb_year:
    # Find the highest rating in the subset
    best_rating = subset['rating'].max()

    # Filter the books with the best rating
    best_books = subset[subset['rating'] == best_rating]

    # Print the year and the best-rated books for that year
    print(f"Year: {year}")
    print(best_books[['name', 'rating']])  # Display book name and rating
    print("-" * 50)  # Separator for readability



Year: -1500
                       name  rating
1397  The Epic of Gilgamesh     3.6
--------------------------------------------------
Year: -800
                      name  rating
245  The Iliad/The Odyssey    4.01
--------------------------------------------------
Year: -560
                name  rating
1396  Aesop's Fables    4.03
--------------------------------------------------
Year: -512
               name  rating
673  The Art of War    3.92
--------------------------------------------------
Year: -500
                  name  rating
745  The Bhagavad Gita    4.06
--------------------------------------------------
Year: -458
              name  rating
1814  The Oresteia    3.96
--------------------------------------------------
Year: -442
         name  rating
776  Antigone    3.52
--------------------------------------------------
Year: -440
               name  rating
2526  The Histories    3.94
--------------------------------------------------
Year: -431
       name  rating
